### PR TITLE
SUBMARINE-277. Support Spark Interpreter add sparkSQL interpreter

### PR DESCRIPTION
**Pull request by hhhizzz**
### What is this PR for?
Add SparkSQL interpreter and improve the structure of interpreter

### What type of PR is it?
Feature and Improvement

### Todos
* [*] - add sparkSQL interpreter
* [*] - add Abstract interpreter

### What is the Jira issue?
https://issues.apache.org/jira/browse/SUBMARINE-277

### How should this be tested?
https://travis-ci.org/hhhizzz/submarine/builds/617065629

### Questions:
* Does the licenses files need an update? No
* Are there breaking changes for older versions? Yes, some change in InterpreterProcess
* Does this needs documentation? No

### CODE REVIEW

File: submarine-workbench/interpreter/python-interpreter/src/main/java/org/apache/submarine/interpreter/PythonInterpreter.java [hunk 1]
```diff
@@ -19,104 +19,58 @@
...
-  @Override
-  public boolean test() {
-    open();
-    String code = "1 + 1";
-    InterpreterResult result = interpret(code);
-    LOG.info("Execution Python Interpreter, Calculation formula {}, Result = {}",
-        code, result.message().get(0).getData());
-
-    if (result.code() != InterpreterResult.Code.SUCCESS) {
+      e.printStackTrace();
```
Comment (liuxunorg) [inline submarine-workbench/interpreter/python-interpreter/src/main/java/org/apache/submarine/interpreter/PythonInterpreter.java]: Need change to `LOG.error(e.getMessage(), e);`

File: submarine-workbench/interpreter/spark-interpreter/src/main/java/org/apache/submarine/interpreter/SparkInterpreter.java [hunk 2]
```diff
@@ -18,158 +18,72 @@
...
+              "|  2|null|\n" +
+              "+---+----+")
+      );
+      close();
+      return success;
     } catch (InterpreterException e) {
-      LOG.error(e.getMessage(), e);
-    } catch (MalformedURLException e) {
-      LOG.error(e.getMessage(), e);
+      e.printStackTrace();
```
Comment (liuxunorg) [inline submarine-workbench/interpreter/spark-interpreter/src/main/java/org/apache/submarine/interpreter/SparkInterpreter.java]: Need change to `LOG.error(e.getMessage(), e);`

File: submarine-workbench/interpreter/spark-interpreter/src/main/java/org/apache/submarine/interpreter/SparkSqlInterpreter.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+                      "))))");
+      sparkInterpreter.interpret("gr.toDF.registerTempTable(\"gr\")");
+
+      InterpreterResult result = interpret("select * from gr");
+      LOG.info("Execution SparkSQL Interpreter, Calculation Spark Code {}, Result =\n {}",
+               result.code(), result.message().get(0).getData()
+      );
+      if (result.code() != InterpreterResult.Code.SUCCESS) {
+        return false;
+      }
```
Comment (liuxunorg) [inline submarine-workbench/interpreter/spark-interpreter/src/main/java/org/apache/submarine/interpreter/SparkSqlInterpreter.java]: Need add sql `select * from gr` execute result check.

File: submarine-workbench/interpreter/spark-interpreter/src/main/java/org/apache/submarine/interpreter/SparkSqlInterpreter.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+      return true;
+    } catch (InterpreterException e) {
+      e.printStackTrace();
```
Comment (liuxunorg) [inline submarine-workbench/interpreter/spark-interpreter/src/main/java/org/apache/submarine/interpreter/SparkSqlInterpreter.java]: Can you change to `LOG.error` ??

File: submarine-workbench/interpreter/spark-interpreter/src/test/java/org/apache/submarine/interpreter/SparkInterpreterTest.java [hunk 5]
```diff
@@ -48,6 +47,8 @@
   @Before
   public void setUp() {
     mockRemoteEventClient = mock(RemoteInterpreterEventClient.class);
+    //SparkInterpreter will change the current thread's classLoader
```
Comment (liuxunorg) [inline submarine-workbench/interpreter/spark-interpreter/src/test/java/org/apache/submarine/interpreter/SparkInterpreterTest.java]: `//SparkInterpreter` need to add a space.

File: submarine-workbench/interpreter/spark-interpreter/src/test/resources/hive-site.xml [hunk 6]
```diff
@@ -0,0 +1,7 @@
+<configuration>
```
Comment (liuxunorg) [inline submarine-workbench/interpreter/spark-interpreter/src/test/resources/hive-site.xml]: When will hive-site.xml be loaded?

File: submarine-workbench/interpreter/interpreter-engine/src/main/java/org/apache/submarine/interpreter/AbstractInterpreter.java [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+ * Accepting thrift connections from Submarine Workbench Server.
+ */
+public abstract class AbstractInterpreter implements Interpreter {
+  private static final Logger LOG = LoggerFactory.getLogger(AbstractInterpreter.class);
+
+  protected org.apache.zeppelin.interpreter.Interpreter zeppelinInterpreter;
+
+  private InterpreterContext interpreterContext;
+
+  private String extractScalaVersion() throws InterpreterException {
```
Comment (liuxunorg) [inline submarine-workbench/interpreter/interpreter-engine/src/main/java/org/apache/submarine/interpreter/AbstractInterpreter.java]: can `extractScalaVersion()` move to `SparkInterpreter` class?

File: submarine-workbench/interpreter/spark-interpreter/README.md [hunk 8]
```diff
@@ -17,13 +17,16 @@
 ## Test Submarine Spark Interpreter
 
 ### Execute test command
-```
-export SUBMARINE_HOME=/path/to/your/submarine_home
-java -jar spark-interpreter-0.3.0-SNAPSHOT-shade.jar spark spark-interpreter-id test
+```bash
+# spark interpreter
+java -jar spark-interpreter-{{version}}-shade.jar spark spark-interpreter-id test
```
Comment (liuxunorg) [inline submarine-workbench/interpreter/spark-interpreter/README.md]: Now, `spark-interpreter-{{version}}-shade.jar` jar package name already change to `submarine-spark-interpreter-{{version}}-shade.jar`

File: submarine-workbench/interpreter/spark-interpreter/src/test/resources/hive-site.xml (hunk 6 above)
Comment (hhhizzz) [reply submarine-workbench/interpreter/spark-interpreter/src/test/resources/hive-site.xml]: the SparkSQL will need hive-site to create a table.

File: submarine-dist/src/assembly/distribution.xml [hunk 9]
```diff
@@ -148,13 +148,6 @@
         <include>submarine-python-interpreter-${project.version}-shade.jar</include>
       </includes>
     </fileSet>
-    <fileSet>
-      <directory>../submarine-workbench/interpreter/spark/scala-2.11</directory>
-      <outputDirectory>/lib/interpreter/spark/scala-2.11</outputDirectory>
-      <includes>
-        <include>spark-scala-2.11-${zeppelin.version}.jar</include>
-      </includes>
-    </fileSet>
```
Comment (liuxunorg) [inline submarine-dist/src/assembly/distribution.xml]: Can `spark-scala-2.11-${zeppelin.version}.jar` be included in `spark-interpreter-${zeppelin.version}-shade.jar` ?

Comment (hhhizzz) [reply submarine-dist/src/assembly/distribution.xml]: Yes, All three Spark interpreters(2.10, 2.11, 2.12) have been packaged into the spark-interpreter-${zeppelin.version}-shade.jar. No more System Variable  like SUBMARINE_HOME or ZEPPELIN_HOME

Comment (liuxunorg) [reply submarine-dist/src/assembly/distribution.xml]: Cool!

**Pull request closed without merging**